frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx

```diff
@@ -0,0 +1,339 @@
+    useState(false)
+  const [hasNoMicPermissions, setHasNoMicPermissions] = useState(false)
+
+  const uploadTheFile = useCallback(
+    (file: File) => {
+      uploadFiles({
+        files: [file],
+        uploadClient,
+        widgetMgr,
+        widgetInfo: element,
+      }).then(({ successfulUploads }) => {
+        const upload = successfulUploads[0]
+        if (upload && upload.fileUrl.deleteUrl) {
+          setDeleteFileUrl(upload.fileUrl.deleteUrl)
+        }
+      })
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [uploadClient, widgetMgr]
+  )
+
+  useEffect(() => {
+    // this first part is to ensure we prompt for getting the user's media devices
+    navigator.mediaDevices
+      .getUserMedia({ audio: true })
+      .then(() => {
+        RecordPlugin.getAvailableAudioDevices().then(devices => {
+          setAvailableAudioDevices(devices)
+          if (devices.length > 0) {
+            setActiveAudioDeviceId(devices[0].deviceId)
+          }
+        })
+      })
+      .catch(_err => {
+        setHasNoMicPermissions(true)
+      })
+  }, [])
+
+  const initializeWaveSurfer = useCallback(() => {
+    if (waveSurferRef.current === null) return
+    if (wavesurfer) wavesurfer.destroy()
+
+    const ws = WaveSurfer.create({
+      container: waveSurferRef.current,
+      waveColor: theme.colors.primary,
+      progressColor: theme.colors.bodyText,
+      height: WAVEFORM_HEIGHT - 2 * WAVEFORM_PADDING,
+      barWidth: BAR_WIDTH,
+      barGap: BAR_GAP,
+      barRadius: BAR_RADIUS,
+      cursorWidth: CURSOR_WIDTH,
+    })
+
+    ws.on("timeupdate", time => {
+      setProgressTime(formatTime(time * 1000)) // get from seconds to milliseconds
+    })
+
+    ws.on("pause", () => {
+      forceRerender()
+    })
+
+    const recordPlugin = ws.registerPlugin(
+      RecordPlugin.create({
+        scrollingWaveform: false,
+        renderRecordedAudio: true,
+      })
+    )
+
+    recordPlugin.on("record-end", blob => {
+      const url = URL.createObjectURL(blob)
+      setRecordingUrl(url)
+
+      const file = new File([blob], "audio.wav", { type: blob.type })
+      uploadTheFile(file)
+
+      ws.setOptions({
+        // This color is hardcoded because I've spent an hour trying to get it to work
+        // with the theme and I'm giving up for now. The complications arise due to the
+        // way wavesurfer is using these colors. Specifically, the progress color "tints" the wave color
+        // and the opacity of this wave color is not being respected, so to make things work for now hardcoding
+        // this color that looks "okay" on both themes.
+        waveColor: "#A5A5AA",
+      })
+    })
+
+    recordPlugin.on("record-progress", time => {
+      setRecordingTime(formatTime(time))
+    })
+
+    setWavesurfer(ws)
+    setRecordPlugin(recordPlugin)
+
+    return () => {
+      if (ws) ws.destroy()
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
**question:** Why are we disabling this? Will it lead to stale values in the callback closure?

Reply:
This was initially a quick fix for a linter issue, but the code in question is no longer needed. I’ll remove both the line and the disable. Thanks for catching that!